### PR TITLE
fix some bug & support PlayAll

### PR DESCRIPTION
**Pull request by Yc7521**

### CODE REVIEW

File: OsuPlayer.Wpf/ViewModels/CollectionPageViewModel.cs [hunk 1]
```diff
@@ -144,10 +144,13 @@ public ICommand DirectPlayCommand
...
                     var beatmap = (BeatmapDataModel)param;
                     var map = _beatmapDbOperator.GetBeatmapByIdentifiable(beatmap);
+                    var maps = Beatmaps.Select(m => _beatmapDbOperator.GetBeatmapByIdentifiable(m)).Where(m => m != null).ToList();
+                    var i = maps.FindIndex(m => m == map);
                     if (map == null) return;
                     await PlayController.Default.PlayNewFile(map);
                     await Services.Get<PlayerList>()
-                        .RefreshPlayListAsync(PlayerList.FreshType.All, PlayListMode.RecentList);
+                        .RefreshPlayListAsync(PlayerList.FreshType.All, PlayListMode.Collection, maps);
+                    Services.Get<PlayerList>().Pointer = i == -1 ? 0 : i;
```
Comment (Milkitic) [inline OsuPlayer.Wpf/ViewModels/CollectionPageViewModel.cs]: Here I made some bugs. My original intention is making `Pointer` property's setter private and call `RefreshPlayListAsync()` to refresh the indexes automatically. It's not recommended to set the Pointer explicitly.

File: OsuPlayer.Common/Player/PlayerList.cs [hunk 2]
```diff
@@ -96,7 +96,7 @@ public int Pointer
                         break;
                 }
 
-            if (force || beatmaps != null || freshType != FreshType.None || Indexes == null || Indexes.Count == 0)
+            if (force || beatmaps != null || freshType != FreshType.None || Indexes == null)
```
Comment (Milkitic) [inline OsuPlayer.Common/Player/PlayerList.cs]: I forgot these uses, but sure it was needed so I added them

File: OsuPlayer.Common/Player/PlayerList.cs [hunk 3]
```diff
@@ -82,7 +82,7 @@ public int Pointer
             }
             if (beatmaps != null)
                 PlayListMode = PlayListMode.Collection;
-            if (force || beatmaps != null || freshType == FreshType.All || Entries.Count == 0)
+            if (force || beatmaps != null || freshType == FreshType.All)
```
Comment (Milkitic) [inline OsuPlayer.Common/Player/PlayerList.cs]: I forgot these uses, but sure it was needed so I added them

**Comment by xyc001** [comment]
Well, refactor the playlist logic. Here are some suggestions.
The `PlayListMode` may be useless.
When the recent playlist need be played, may can get the playlist from database and set the current playlist.
When a song item be clicked, may need set the current playlist to be this collection or recent playlist.
The current playlist and recent playlist may be stored in the database.
The `RefreshPlayListAsync` can set an offset for playing. (means `this.Pointer=offset`)

**Comment by Milkitic** [comment]
> Well, refactor the playlist logic. Here are some suggestions.
> The `PlayListMode` may be useless.
> When the recent playlist need be played, may can get the playlist from database and set the current playlist.
> When a song item be clicked, may need set the current playlist to be this collection or recent playlist.
> The current playlist and recent playlist may be stored in the database.
> The `RefreshPlayListAsync` can set an offset for playing. (means `this.Pointer=offset`)

Thanks for suggestion, I got you. Btw can Yc7521 you make changes only on which I didn't mentioned in order to merge?

**Comment by Yc7521** [comment]
It is you who chooses code to merge into the master.

**Pull request merged**